_posts/2021-05-06-pixel-tests.md

```diff
@@ -0,0 +1,191 @@
+---
+title: Testing all the pixels
+author: mvo
+date: 2021-05-06 10:00
+tags: cockpit
+slug: pixel-testing
+category: tutorial
+comments: 'true'
+---
+
+The Cockpit integration tests can now contain "pixel tests". Such a
+test will take a screenshot with the browser and compare it with a
+reference.  The idea is that we can catch visual regressions much
+easier this way than if we would hunt for them in a purely manual
+fashion.
+
+## Preparing a repository for pixel tests
+
+A pixel test will take a screenshot of part of the Cockpit UI and
+compare it with a reference.  Thus, these reference images are
+important and play the biggest role.
+
+A large part of dealing with pixel tests will consequently consist of
+maintaining the reference images.  At the same time, we don't want to
+clog up our main source repository with them.  While the number and
+size of the reference images at any one point in time should not pose
+a problem, we will over time accumulate a history of them that we are
+afraid would dominate the source repository.
+
+Thus, the reference images are not stored in the source repository.
+Instead, we store them in an external repository that is linked into
+the source repository as a submodule.  That external repository
+doesn't keep any history and can be aggressively pruned.
```

Comment:
Possibly link to https://git-scm.com/docs/git-submodule and/or https://git-scm.com/book/en/v2/Git-Tools-Submodules ?

We don't necessarily need to explain it here, but what happens if we prune the external repo and check out an old version of cockpit.git? I.e. the referenced SHA does not exist? Or does the mere existence of a SHA reference in cockpit.git count keep all the SHAs in the github repo alive? In the latter case, the focus would shift from "don't accumulate cruft on github.com" to "we don't want to check out the entire image history when cloning cockpit.git".

Reply:
> We don't necessarily need to explain it here, but what happens if we prune the external repo and check out an old version of cockpit.git?

The build will likely fail where it calls `pixel-tests pull`.  If we get past that, each assert_pixels call will fail because the expected reference image isn't there.

This is not great... should we add a explicit knob to turn off pixel tests? Like TEST_PIXELS=0 in the envrionment?

Reply:
> Possibly link 

Yes!